sympy/matrices/expressions/blockmatrix.py

```diff
@@ -197,10 +197,8 @@ def _eval_adjoint(self):
 
     def _eval_trace(self):
         if self.rowblocksizes == self.colblocksizes:
-            return Add(*[trace(self.blocks[i, i])
-                        for i in range(self.blockshape[0])])
-        raise NotImplementedError(
-            "Can't perform trace of irregular blockshape")
+            blocks = [self.blocks[i, i] for i in range(self.blockshape[0])]
```

Comment:
So the idea is to let the error arise from the place that has an issue, not to catch the issue higher up if all you are going to do is raise an error, right?

Reply:
The calling code previously checked for AttributeError and NotImplementedError. I have changed it so that all relevant types have an `_eval_trace` method (no `AttributeError` any more) and also the methods return `None` rather than raising `NotImplementedError`.